### PR TITLE
Introduce a function context

### PR DESCRIPTION
**Pull request by zhassan-aws**
This is a refactoring PR that introduces a function context, and moves all the codegen functions from the Boogie context to the new function context.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/codegen_boogie/context/boogie_ctx.rs [hunk 1]
```diff
@@ -324,47 +324,35 @@ impl<'tcx> BoogieCtx<'tcx> {
...
 
-    /// Write the program to the given writer
-    pub fn write<T: Write>(&self, writer: &mut T) -> std::io::Result<()> {
-        self.program.write_to(writer)?;
-        Ok(())
-    }
-
-    fn operand_ty(&self, local_decls: &LocalDecls<'tcx>, o: &Operand<'tcx>) -> Ty<'tcx> {
+    fn operand_ty(&self, o: &Operand<'tcx>) -> Ty<'tcx> {
         // TODO: monomorphize
```
Comment (celinval) [inline kani-compiler/src/codegen_boogie/context/boogie_ctx.rs:328]: if you move to StableMIR you won't need to monomorphize. :smile:

File: kani-compiler/src/codegen_boogie/context/kani_intrinsic.rs [hunk 2]
```diff
@@ -26,23 +26,25 @@ pub enum KaniIntrinsic {
...
-        let intrinsics = KaniIntrinsic::VARIANTS;
-        for intrinsic in intrinsics {
-            let attr_sym = rustc_span::symbol::Symbol::intern(intrinsic);
-            if let Some(attr_id) = self.tcx.all_diagnostic_items(()).name_to_id.get(&attr_sym) {
-                if instance.def.def_id() == *attr_id {
-                    debug!("matched: {:?} {:?}", attr_id, attr_sym);
-                    return Some(KaniIntrinsic::from_str(intrinsic).unwrap());
-                }
+/// If provided function is a Kani intrinsic (e.g. assert, assume, cover), returns it
+// TODO: move this function up to `kani_middle` along with the enum
```
Comment (celinval) [inline kani-compiler/src/codegen_boogie/context/kani_intrinsic.rs:30]: +1

**Review by celinval — APPROVED** [review]

File: kani-compiler/src/codegen_boogie/context/boogie_ctx.rs (hunk 1 above)
Comment (zhassan-aws) [reply kani-compiler/src/codegen_boogie/context/boogie_ctx.rs:328]: Nice! I need to educate myself on StableMIR!

File: kani-compiler/src/codegen_boogie/context/kani_intrinsic.rs (hunk 2 above)
Comment (zhassan-aws) [reply kani-compiler/src/codegen_boogie/context/kani_intrinsic.rs:30]: I'll try to do that as part of merging the Boogie backend to main.

**Pull request merged**